feat(graphql): enabling graphql for data platform instance aspects

sgomezvillamor opened the pull request:
This continues the work started here https://github.com/datahub-project/datahub/pull/5728 and so it enables Data Platform Instance aspects being managed via GraphQL.

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

sgomezvillamor commented:
I've been following instructions here https://datahubproject.io/docs/datahub-graphql-core/#extending-the-graph

Although this PR is about adding the aspects to the GraphQL interface, I would like to check the Data Platform Instance entity is searchable. While the instructions mention this related to that:

> In order to enable searching an Entity, you'll need to modify the [SearchAcrossEntities.java](https://github.com/datahub-project/datahub/blob/master/datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/search/SearchAcrossEntitiesResolver.java) resolver, which enables unified search across all DataHub entities.

I couldn't find where to add the DataPlatformInstance, so I cannot confirm in that code whether it is searchable or not. Likely the code has evolved and those instructions are not up-to-date.

chriscollins3456 commented:
hey @sgomezvillamor thanks for the PR! gonna give this a look through now :)

chriscollins3456 commented on datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/types/dataplatforminstance/DataPlatformInstanceType.java:
```diff
@@ -24,7 +24,15 @@
+        Constants.DEPRECATION_ASPECT_NAME,
+        Constants.OWNERSHIP_ASPECT_NAME,
+        Constants.INSTITUTIONAL_MEMORY_ASPECT_NAME,
+        Constants.GLOBAL_TAGS_ASPECT_NAME,
```
what do you think about adding `GLOSSARY_TERMS_ASPECT_NAME` and `DOMAINS_ASPECT_NAME` if the goal is to get this to have an entity profile page? we'd have to update entity-registry.yml to include those aspects for this entity as well

sgomezvillamor replied:
I mentioned that in the previous PR, the one enabling the ingestion for all Data Platform Instance aspects https://github.com/datahub-project/datahub/pull/5728

> I have explicitely skipped GlossaryTerm and Domain because those look more related to data than infrastructure

WDYT? I would skip these until we have clear use cases for that. Adding them after this will be straightforward.

chriscollins3456 replied:
that is totally good with me! it makes sense and is definitely something we can come back to later.

chriscollins3456 commented on datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/types/dataplatforminstance/mappers/DataPlatformInstanceMapper.java:
```diff
@@ -33,6 +47,39 @@ public DataPlatformInstance apply(@Nonnull final EntityResponse entityResponse)
     MappingHelper<DataPlatformInstance> mappingHelper = new MappingHelper<>(aspects, result);
     mappingHelper.mapToResult(DATA_PLATFORM_INSTANCE_KEY_ASPECT_NAME, this::mapDataPlatformInstanceKey);
 
+    final EnvelopedAspect envelopedDataPlatformInstanceProperties = aspects.get(Constants.DATA_PLATFORM_INSTANCE_PROPERTIES_ASPECT_NAME);
+    if (envelopedDataPlatformInstanceProperties != null) {
+      result.setProperties(mapDataPlatformInstanceProperties(
+              new DataPlatformInstanceProperties(envelopedDataPlatformInstanceProperties.getValue().data()), entityUrn)
+      );
+    }
+
+    final EnvelopedAspect envelopedOwnership = aspects.get(Constants.OWNERSHIP_ASPECT_NAME);
+    if (envelopedOwnership != null) {
+      result.setOwnership(OwnershipMapper.map(new Ownership(envelopedOwnership.getValue().data()), entityUrn));
+    }
+
+    final EnvelopedAspect envelopedTags = aspects.get(Constants.GLOBAL_TAGS_ASPECT_NAME);
+    if (envelopedTags != null) {
+      com.linkedin.datahub.graphql.generated.GlobalTags globalTags = GlobalTagsMapper.map(new GlobalTags(envelopedTags.getValue().data()), entityUrn);
+      result.setTags(globalTags);
+    }
+
+    final EnvelopedAspect envelopedInstitutionalMemory = aspects.get(Constants.INSTITUTIONAL_MEMORY_ASPECT_NAME);
+    if (envelopedInstitutionalMemory != null) {
+      result.setInstitutionalMemory(InstitutionalMemoryMapper.map(new InstitutionalMemory(envelopedInstitutionalMemory.getValue().data())));
+    }
+
+    final EnvelopedAspect statusAspect = aspects.get(Constants.STATUS_ASPECT_NAME);
+    if (statusAspect != null) {
+      result.setStatus(StatusMapper.map(new Status(statusAspect.getValue().data())));
+    }
+
+    final EnvelopedAspect envelopedDeprecation = aspects.get(Constants.DEPRECATION_ASPECT_NAME);
+    if (envelopedDeprecation != null) {
+      result.setDeprecation(DeprecationMapper.map(new Deprecation(envelopedDeprecation.getValue().data())));
+    }
+
```
code style: would you mind updating these blocks to use the `mappingHelper` like we do for the key aspect above? a good example to follow would be `DatasetMapper.java` - using the `mappingHelper` to do this aspect check and then either setting the value or calling a mapping function like `mapDataPlatformInstanceKey` below

sgomezvillamor replied:
I took `ContainerMapper.java` as example, as pointed out in the instructions.
I will have a look at `DatasetMapper.java` instead.

sgomezvillamor replied:
addressed in 4730adbe7

chriscollins3456 commented on datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/types/dataplatforminstance/mappers/DataPlatformInstanceMapper.java:
```diff
@@ -33,6 +47,39 @@ public DataPlatformInstance apply(@Nonnull final EntityResponse entityResponse)
+    final EnvelopedAspect envelopedTags = aspects.get(Constants.GLOBAL_TAGS_ASPECT_NAME);
+    if (envelopedTags != null) {
+      com.linkedin.datahub.graphql.generated.GlobalTags globalTags = GlobalTagsMapper.map(new GlobalTags(envelopedTags.getValue().data()), entityUrn);
+      result.setTags(globalTags);
```
while `globalTags` is deprecated, I'm not positive we've migrated everywhere off of the frontend from using it. could you do `result.setGlobalTags` here as well as `result.setTags`?

sgomezvillamor replied:
There is no `result.setGlobalTags` in the `com.linkedin.datahub.graphql.generated.DataPlatformInstance`. 

In https://github.com/datahub-project/datahub/pull/5728 I added the `GlobalTags` aspect in the `entity-registry.yml`, and skipped the deprecated one for `Tags`. With that, the graphql generator produces the following methods about tags:

```
    /**
     * Tags used for searching the data platform instance
     */
    public GlobalTags getTags() {
        return tags;
    }
    /**
     * Tags used for searching the data platform instance
     */
    public void setTags(GlobalTags tags) {
        this.tags = tags;
    }
```

So there is no `setGlobalTags` and instead the `setTags` is about `GlobalTags`.

Do you still see any issue on this?

chriscollins3456 replied:
oh nice! okay that all makes sense and is good to go then.

chriscollins3456 commented on datahub-graphql-core/src/main/resources/entity.graphql:
```diff
@@ -240,6 +240,11 @@ type Mutation {
+    """
+    Update the metadata about a particular DataPlatformInstance
+    """
+    updateDataPlatformInstance(urn: String!, input: DataPlatformInstanceUpdateInput!): DataPlatformInstance
```
I would wait to add this until we actually build out the endpoint/resolver for updating a data platform instance. Otherwise someone might see this as an option to query and it would error out

sgomezvillamor replied:
I'm ok with that.
The main goal of this PR is to make DataPlatformInstance metadata accessible via GraphQL. Update is not a requirement at this moment.

Addressed in f646ed36d

chriscollins3456 replied:
> * The one about using the _mapper helper_ is still pending; I will have a look at some moment during this week.

@sgomezvillamor sounds good! let me know if you have any questions - I'm happy to provide some more examples or collaborate some more on this.

Also the changes to the mapper aren't extremely important, just more code style as we try to move everything to use this mapping helper.

chriscollins3456 commented on datahub-graphql-core/src/main/resources/entity.graphql:
```diff
@@ -1839,6 +1870,36 @@ type DataPlatformInstance implements Entity {
+    """
+    Tags used for searching the data platform instance
+    """
+    tags: GlobalTags
```
if we add domain and glossary terms we can also add that here!

sgomezvillamor replied:
See comment https://github.com/datahub-project/datahub/pull/7177#discussion_r1092361062

chriscollins3456 commented on datahub-graphql-core/src/main/resources/entity.graphql:
```diff
@@ -4040,6 +4101,31 @@ input TagUpdateInput {
     ownership: OwnershipUpdate
 }
 
+"""
+Arguments provided to update a DataPlatformInstance Entity
+"""
+input DataPlatformInstanceUpdateInput {
+    """
+    Update to ownership
+    """
+    ownership: OwnershipUpdate
+
+    """
+    Update to deprecation status
+    """
+    deprecation: DatasetDeprecationUpdate
+
+    """
+    Update to institutional memory, ie documentation
+    """
+    institutionalMemory: InstitutionalMemoryUpdate
+
+    """
+    Update to tags
+    """
+    tags: GlobalTagsUpdate
+}
+
```
Same with holding on this until we build out the update endpoint and add `update` to the `DataPlatformInstanceType`

also for what it's worth, we won't actually be calling a new update endpoint from the UI for any of these fields as they all have their own specific endpoint (ie. `addOwner`, `removeOwner`, `updateDeprecation`, `addLink`, etc.)

But having this exposed via graphql API as an option still makes sense

sgomezvillamor replied:
I'm ok with that, addressed in f646ed36d

chriscollins3456 reviewed (commented):
Nice! this is looking great. I have a few suggestions / changes but once that's settled I think this is ready to rumble!

sgomezvillamor commented:
Thanks @chriscollins3456 for the review.

Here my update:

- I addressed most of the comments, the easy ones :-) 
- The one about using the _mapper helper_ is still pending; I will have a look at some moment during this week.
- I also shared some feedback about why I skipped adding domains and glossary terms.

chriscollins3456 commented:
> * The one about using the _mapper helper_ is still pending; I will have a look at some moment during this week.

@sgomezvillamor sounds good! let me know if you have any questions - I'm happy to provide some more examples or collaborate some more on this.

sgomezvillamor commented:
Hi @chriscollins3456 I have addressed the comment about the code style with the mapperHelper. May you have a look? Let me know if there is any other update that you would like to include in this PR.

From my side, the question raised here https://github.com/datahub-project/datahub/pull/7177#issuecomment-1408960200 is still open to me. With the updates done here in this PR, is the Data Platform Instance searchable by its aspects?

Thanks!

chriscollins3456 commented:
> From my side, the question raised here [#7177 (comment)](https://github.com/datahub-project/datahub/pull/7177#issuecomment-1408960200) is still open to me. With the updates done here in this PR, is the Data Platform Instance searchable by its aspects?

Sorry I didn't even notice that question above! yeah so what you need to do is update `SEARCHABLE_ENTITY_TYPES` which you will see used in the resolver you mention above. You can also add data platform instance type to `AUTO_COMPLETE_ENTITY_TYPES` right below it. You'll know what pieces of the entity's aspects are actually searchable by seeing what has a `@Searchable` annotation above it. so it looks like the `name` and `description` from `dataPlatformInstanceProperties` are searchable already

chriscollins3456 reviewed (approved):
@sgomezvillamor  This is looking great for what you have! I know you might want to make one last change and add to `SEARCHABLE_ENTITY_TYPES` and `AUTO_COMPLETE_ENTITY_TYPES` as well, just let me know when you push that up and we can get this guy in there

sgomezvillamor commented:
Hi @chriscollins3456 ... here it is, last but not least https://github.com/datahub-project/datahub/pull/7177/commits/f975307b28b1e4ffa0da80f878f080031d36ab0b :-)

chriscollins3456 reviewed (changes requested):
@sgomezvillamor actually this last commit does cause concern..

If we add it to `SEARCHABLE_ENTITY_TYPES` and `AUTO_COMPLETE_ENTITY_TYPES` then data platform instances will start showing up in search and auto-complete results.

however, we don't have search result cards or a profile page for data platform instances built on the frontend yet, so I think this will definitely cause problems.

so before we add those two things in `SearchUtils.java` we need to build out the frontend pieces for data platform instances!

sgomezvillamor commented:
Good point on search and autocompletion @chriscollins3456 ! I have reverted that commit and explicitly mentioned in the PR description the features not covered in the scope of this PR (updates, search+autocomplete, UI elements).

chriscollins3456 reviewed (approved):
alright nice! thanks so much for this. Excited for this feature development to keep on rolling!

sgomezvillamor commented:
Thanks to you for your speedy and responsive support!

sgomezvillamor commented:
By the way, I cannot merge, so all yours, folks!
